File: 206-reverse-linked-list/206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode reverseMyList(ListNode head){
        if(head==null || head.next==null) return head;
        
        ListNode newHead=reverseMyList(head.next);
    
        head.next.next=head;
        head.next=null;
        return newHead;
    }
    
    public ListNode reverseList(ListNode head) {
            return reverseMyList(head);
    }
}

File: 206-reverse-linked-list/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Sol 1 Recursive
# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         if not head or not head.next:
#             return head
#         newHead = self.reverseList(head.next)
#         head.next.next = head
#         head.next = None
#         return newHead

# Sol 2 Iterative
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        while head:
            curr=head
            head=head.next
            curr.next=prev
            prev=curr
        return prev
            
            
        